core/sched.cc

```diff
@@ -380,7 +405,10 @@ void cpu::reschedule_from_interrupt(bool called_from_yield,
         n->_runtime.add_context_switch_penalty();
     }
     preemption_timer.cancel();
-    if (n->_realtime._priority == 0) {
+    if (n->_realtime.has_slice()) {
+        assert(n->_realtime.has_remaining());
+        preemption_timer.set_with_irq_disabled(now + n->_realtime.remaining());
+    } else if (n->_realtime._priority == 0) {
```

Comment:
I would reverse the order if() to check realtime.priority==0 first and the has_slice() will be the else if. This is the most typical case, so it's more efficient to check it first.

Also, as I noted above twice already, I think if priority.realtime==0 we shouldn't even check the has_slice() which is meaningless - which also means you need to reverse the order of the if and else - not just for performance but also for correctness.